src/librustc/traits/select.rs

```diff
@@ -3380,21 +3396,20 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                 // with what our behavior should be there. -nikomatsakis
                 let InferOk { obligations, .. } = self.infcx
                     .at(&obligation.cause, obligation.param_env)
-                    .eq(target, source_trait) // FIXME -- see below
+                    .eq(target, source_ty) // FIXME -- see above
                     .map_err(|_| Unimplemented)?;
                 nested.extend(obligations);
+                */
 
-                // Register one obligation for 'a: 'b.
-                let cause = ObligationCause::new(
-                    obligation.cause.span,
-                    obligation.cause.body_id,
-                    ObjectCastObligation(target),
+                // Register an obligation for `dyn TraitA: TraitB`.
+                nested.extend(
+                    data_b.iter()
+                        .map(|d| predicate_to_obligation(d.with_self_ty(tcx, source_ty))),
```

Comment:
Maybe this gives you a `Prediate::Trait`, when you want a `Predicate::Subtype`? I don't know this code very well, I'm not sure I can be of much help here.

Reply:
I already showed this bit to eddyb a few days ago, and he thought it was right... but you make a fair point; I'll check that out.

Reply:
Well, we're dealing with an `ExistentialPredicate` here in fact, so there's no `Subtrait`... might have to wait for eddyb to review. Thanks thtough.

Reply:
@oli-obk Interesting: having this new code plus the old code (uncommented) gives the same error!